### PR TITLE
Support providing Jars with alternative Java compiler implementations

### PR DESCRIPTION
**Pull request by jjohannes**
### Context
Proposal to provide a solution for: #15942

The proposal is to allow the definition of a `customCompilerClasspath` of the `JavaCompile` task. If the `customCompilerClasspath` contains an implementation of `javax.tools.JavaCompiler` that can be discovered through Java's `ServiceLoader` it is used in place of the default JDK compiler (JavacTool) – "on top" of the selected Java toolchain. 

This is how you would configure it to use the Eclipse Compiler (ECJ):

```kotlin
val ecj = configurations.create("ecj")
dependencies {
    "ecj"("org.eclipse.jdt:ecj:3.33.0")
}

tasks.withType(JavaCompile::class).configureEach {
    customCompilerClasspath.from(ecj)
    options.headerOutputDirectory.convention(null) // see #12904
}
```

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation

### CODE REVIEW

**Comment by eskatos** [comment]
@bot-gradle test this

**Comment by eskatos** [comment]
@bot-gradle test this

**Comment by jjohannes** [comment]
The PR does not show my latest changes on the branch. I think GH is having issues right now. If CI fails, pls try again later.

**Comment by eskatos** [comment]
@bot-gradle test this

**Comment by eskatos** [comment]
@bot-gradle test RFN

**Comment by eskatos** [comment]
@bot-gradle test RFN

File: subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/JdkTools.java [hunk 1]
```diff
@@ -103,7 +108,18 @@ private JavaCompiler buildJavaCompiler() {
             if (isJava9Compatible) {
                 clazz = isolatedToolsLoader.loadClass("javax.tools.ToolProvider");
                 try {
-                    JavaCompiler compiler = (JavaCompiler) clazz.getDeclaredMethod("getSystemJavaCompiler").invoke(null);
+                    JavaCompiler compiler = null;
+                    for (JavaCompiler compilerCandidate : ServiceLoader.load(JavaCompiler.class, isolatedToolsLoader)) {
+                        if (!compilerCandidate.getClass().getName().equals(DEFAULT_COMPILER_IMPL_NAME)) {
+                            // take the first compiler found on the custom compiler classpath that is not the JavacTool system compiler (if any)
+                            compiler = compilerCandidate;
```
Comment (tresat) [inline subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/JdkTools.java:115]: I think this should probably at least log which compiler is selected to debug, to enable troubleshooting if multiple compilers are accidentally included.

File: subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/reflect/GradleStandardJavaFileManager.java [hunk 2]
```diff
@@ -18,18 +18,22 @@
...
 import java.io.IOException;
 import java.net.URLClassLoader;
+import java.nio.file.Path;
+import java.util.Collection;
 import java.util.Set;
 
 import static org.gradle.api.internal.tasks.compile.filter.AnnotationProcessorFilter.getFilteredClassLoader;
 
-public class GradleStandardJavaFileManager extends ForwardingJavaFileManager<StandardJavaFileManager> {
+public class GradleStandardJavaFileManager extends ForwardingJavaFileManager<StandardJavaFileManager> implements StandardJavaFileManager {
```
Comment (tresat) [inline subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/reflect/GradleStandardJavaFileManager.java:36]: What's the reason for this change?

**Review by tresat — APPROVED** [review]

File: subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/JavaHomeBasedJavaCompilerFactory.java [hunk 3]
```diff
@@ -27,7 +27,7 @@
 import java.util.concurrent.ConcurrentHashMap;
 
 public class JavaHomeBasedJavaCompilerFactory implements Factory<JavaCompiler>, Serializable {
-    private final List<File> compilerPluginsClasspath;
+    final List<File> compilerPluginsClasspath;
```
[the remark this answers is not in the document]
Comment (jjohannes) [reply subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/JavaHomeBasedJavaCompilerFactory.java:30]: Turned into a _protected_ getter for consistency with how things are done elsewhere in the code base.

**Comment by jjohannes** [comment]
Thank you @tresat for the review. I addressed the comments.

File: subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/reflect/GradleStandardJavaFileManager.java (hunk 2 above)
Comment (tresat) [reply subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/reflect/GradleStandardJavaFileManager.java:36]: Got it, that makes sense.

**Comment by TheMrMilchmann** [comment]
My initial testing (in the context of https://github.com/TheMrMilchmann/gradle-ecj/issues/13) seemed to confirm that this is working perfectly fine and using the toolchains as expected. I'll get back to do some more exhaustive testing at a later date but don't see any issue currently.

@jjohannes In the meantime, I'm curious if you have an idea of how to verify that the correct compiler has been used in a functional test. Previously, I inspected the logs but with the new API, there does not seem to be a unique output that I could check for. Would logging the used compiler implementation be possible?

**Comment by jjohannes** [comment]
That's great 👍. Thanks @TheMrMilchmann.

Which debug output did you use in the other solution? Was that info or debug level?

I added logging to this PR in the last change. You probably did not have that in your version yet. https://github.com/gradle/gradle/pull/24649/commits/fef7ea239ef75b0a2e895ca8ef96d06939ff5b50#diff-6a9ca4aa877127a4e8b891057fb29a41545c2ad8e95a5a21f7d052245c11f752R118

I was wondering if tat should be on info level.

**Comment by TheMrMilchmann** [comment]
> I added logging to this PR in the last change. You probably did not have that in your version yet. https://github.com/gradle/gradle/commit/fef7ea239ef75b0a2e895ca8ef96d06939ff5b50#diff-6a9ca4aa877127a4e8b891057fb29a41545c2ad8e95a5a21f7d052245c11f752R118

You're absolutely right, I forgot to check the date of the build. This looks exactly like what I was hoping for.

> Which debug output did you use in the other solution? Was that info or debug level?

Currently, I'm using the following assertion to verify that `java.exe` is used for compilation (instead of `javac`):

```java
assertTrue(buildResult.output.contains("Compiling with Java command line compiler '.+[/\\\\]bin[/\\\\]java(.exe)?'.".toRegex()))
```

This is not ideal but if the compilation succeeds, I know that ECJ must have been used. The output is currently captured with `--info` logging, but I don't really have a strong opinion on whether this should be logged at info or debug level. From my perspective, the worst thing that could happen is that my tests might run a bit slower. 😅

**Comment by jjohannes** [comment]
I would tend to move the added logging to `--info`. It feels like this is information on the level as the other logging in this area – similiar to `Compiling with Java command line compiler...`.

If you agree @tresat, I can move it to `--info`.

**Comment by tresat** [comment]
Sounds good to me.  

Some others are going to take a look at this soon and provide further guidance.

**Comment by adammurdoch** [comment]
I think a better approach would be to (continue to) detangle the API for configuration and initialization of tool chains from the tasks that use the toolchains, rather than add more of these kinds of methods to specific tasks. So, I would add a new factory method to, say,`ToolchainRegistry` that given an implementation classspath creates a `Toolchain` instance, or perhaps better a`Provider<Toolchain>` instance. This instance can then be wired into whichever components or tasks need it.

**Comment by jjohannes** [comment]
In general, I agree that

>  detangle the API for configuration and initialization of tool chains from the tasks

could be a better approach.

I however have not enough context/knowledge to know what should be done concretely with the current state of the `org.gradle.jvm.toolchain.*` interfaces. I assume you refer to these when you talk about `ToolchainRegistry`? (The interface `ToolchainRegistry` is somewhere else and, IIUC, refers to the "old" Toolchain concept.)

IIUC, the only place users can provide customizations for Java toolchains right now is `JavaToolchainResolver`, in which the users may construct `JavaToolchainDownload` instances. I am not sure where the implementation classpath / `Provider<Toolchain>` would fit in here.

If someone could give guidance, I am happy to change the implementation here accordingly.

**Comment by jjohannes** [comment]
A concrete thing I could imagine is to make the `customCompilerClasspath` part of `JavaCompile.getJavaCompiler()` input property (which is of type `org.gradle.jvm.toolchain.JavaCompiler`). And then also make defining the "Compiler Implementation" part of `org.gradle.jvm.toolchain.JavaToolchainSpec`.

Then we would need a ways to specify the "Compiler Implementation" as part of the Java toolchain registry. Maybe making it part as an optional additional download URI of `JavaToolchainDownload` (?).

This feels like a lot. What do you think about adding the implementation proposed in this PR as a first step (possibly keeping it on `@Incubating` for longer/ever) and replace it later with something integrated into the toolchain registry when work on toolchains in general is picked up again?

**Review by lkasso — APPROVED** [review]

**Comment by jjohannes** [comment]
Gradle 8.4 + this PR: https://github.com/jjohannes/gradle/tree/release-8.4-ecj-patch/gradle-bin

**Comment by AceTheFace** [comment]
Is there any ETA when this will be merged? Will it make it into 8.6 release?

**Comment by big-guy** [comment]
Sorry, we don't want to model things this way.  We know that supporting other Java compilers would be a good thing to have, but we don't want to expose this as another configuration with arbitrary jars that we execute. 

How exactly this needs to be done is unclear, but we think it would be more closely tied to a Java toolchain that provides a different Java compiler. 

We'll leave the original issue open, but close this PR.

**Comment by AceTheFace** [comment]
:-( Problem is, that this functionality was available with Gradle 6.x and has been dropped with 7 - which now forces us to use unstable internal API, which might change in every release. So any official way to support this would highly be appreciated.

**Comment by jjohannes** [comment]
>  forces us to use unstable internal API

Our problem is that there are no APIs (internal or public) at the moment to get the custom compiler used by the compile worker JVMs. And we need that to get acceptable compile performance in a (very) large project.

I think no matter where this topic goes next, this PR contains a valuable technical solution to get the additional compiler Jar into the worker. So no matter how a public API surface may look one day, I think something like that has to be added in any case.

@big-guy what would be the concrete next steps? There are folks from the community (@TheMrMilchmann, me and others) who are ready to invest time and resources to make this happen. What can we do? As I wrote already, an initial internal only API that allows us to get compiler Jars into the worker, would be very helpful already. Then we have a technical foundation to built upon (in custom plugins/builds). But without this addition, we cannot solve this without patching Gradle.
